### PR TITLE
fix(auth): Fix for when loading credentials the success/error is fired twice

### PR DESCRIPTION
**Pull request by gpanshu**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:* Fix for when loading credentials the success/error is fired twice it returns an already resumed error in the continuation

*How did you test these changes?*
(Please add a line here how the changes were tested)

- [ ] Added Unit Tests
- [ ] Added Integration Tests

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt [hunk 1]
```diff
@@ -56,9 +56,9 @@ internal class CredentialStoreClient(configuration: AuthConfiguration, context:
     ) {
         var capturedSuccess: Result<AmplifyCredential>? = null
         var capturedError: Exception? = null
-        val token = StateChangeListenerToken()
+        var token: StateChangeListenerToken? = StateChangeListenerToken()
         credentialStoreStateMachine.listen(
-            token,
+            token as StateChangeListenerToken,
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt:61]: Maybe wrap the token call in a `?.let` so you don't have to do a force cast.

**Review by tylerjroach — COMMENTED** [review]
I'd still like for us to try and create a rapid fire test that replicates the occasional failure here as this fix is masking the problem a bit. credentialStoreStateMachine.cancel() immediately cancels the token through a pending cancellations set.

The race condition here appears that maybe 2 results are fired in `listen` from multiple threads, and since there was no further check for if token was subscribed, the fire could happen twice.

Not sure if this is the right fix though. Other options could be setting a lock on listen callbacks (also feels a bit hacky), as the issue is most likely inside the structuring/threading of CredentialStoreStateMachine. Want to see what others think on this.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt (hunk 1 above)
Comment (mattcreaser) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt:61]: Or a simple `!!`

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt [hunk 2]
```diff
@@ -72,11 +72,12 @@ internal class CredentialStoreClient(configuration: AuthConfiguration, context:
...
                         val error = capturedError
-                        if (success != null) {
-                            credentialStoreStateMachine.cancel(token)
+                        if (success != null && token != null) {
+                            credentialStoreStateMachine.cancel(token as StateChangeListenerToken)
+                            token = null
                             onSuccess(success)
-                        } else if (error != null) {
-                            credentialStoreStateMachine.cancel(token)
+                        } else if (error != null && token != null) {
```
Comment (mattcreaser) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt:79]: Token also has to be set to null for the error case.

**Review by mattcreaser — CHANGES_REQUESTED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt (hunk 1 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt:61]: Done

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt (hunk 2 above)
Comment (gpanshu) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/CredentialStoreClient.kt:78]: Good catch. Done.

**Review by tylerjroach — DISMISSED** [review]

**Review by sdhuka — APPROVED** [review]

**Review by sdhuka — APPROVED** [review]

File: aws-analytics-pinpoint/src/androidTest/java/com/amplifyframework/analytics/pinpoint/PinpointAnalyticsInstrumentationTest.kt [hunk 3]
```diff
@@ -53,11 +54,18 @@ class PinpointAnalyticsInstrumentationTest {
         val context = ApplicationProvider.getApplicationContext<Context>()
         @RawRes val resourceId = Resources.getRawResourceId(context, CREDENTIALS_RESOURCE_NAME)
         val userAndPasswordPair = readCredentialsFromResource(context, resourceId)
-        synchronousAuth.signOut()
-        synchronousAuth.signIn(
-            userAndPasswordPair!!.first,
-            userAndPasswordPair.second
-        )
+        try {
```
Comment (tylerjroach) [inline aws-analytics-pinpoint/src/androidTest/java/com/amplifyframework/analytics/pinpoint/PinpointAnalyticsInstrumentationTest.kt:57]: We should not be leaking state between tests.

**Review by tylerjroach — CHANGES_REQUESTED** [review]

File: aws-analytics-pinpoint/src/androidTest/java/com/amplifyframework/analytics/pinpoint/PinpointAnalyticsInstrumentationTest.kt (hunk 3 above)
Comment (gpanshu) [reply aws-analytics-pinpoint/src/androidTest/java/com/amplifyframework/analytics/pinpoint/PinpointAnalyticsInstrumentationTest.kt]: reverted as it passes locally without this change

**Review by tylerjroach — APPROVED** [review]

**Review by sdhuka — APPROVED** [review]

**Pull request merged**